### ISSUE TITLE
Populate values in filter builder

### ISSUE THREAD

**Issue by buremba** [rakam-io/rakam-ui#65]
We also want to have an ability where users can see possible values against a column when a column is selected from filter dropdown. For example if i select "os" as one column from filter dropdown and the select operator as "equals" and then in value text field, if we can show all distinct possible values for those columns that people can chose from so that people will not have to explicitly write or remember those values. For example "os" could have possible values as "android 1.0", "android 2.0", "iOS 6.0", "iOS 7.0" etc. This probably can be achieved by materialising a collection_filter table where we can maintain distinct values for selected columns. This tables needs to be populated (insert overwrite every time new data is reloaded). This process of materializing this collection_filter table should happen in a scheduled manner or on every data load time. We can take either approach, if we make it scheduled based then it becomes automatic and no dependency on when data is getting pushed.

**Comment by sambitdixit** [comment]
Any update when this feature can be picked up.

**Comment by buremba** [comment]
Rakam is real-time so `INSERT OVERWRITE` approach doesn't work for us. There are two approaches that I can think of:

1. We can dynamically issue `SELECT DISTINCT field_name FROM collection WHERE _time BETWEEN autosuggestion_period_start and autosuggestion_period_start LIMIT 10`. It doesn't require materialization and for big data-sets, we can basically use `TABLE SAMPLE` feature in order to be make sure that latency is not high.

2. Incremental materialization. We can create an incremental materializated view `SELECT field_name, max(_time), min(_time) FROM collection GROUP BY 1` for each field and use it in order to perform lookups. We need `count of fields * count of collections` materialized views.

1 seems to be a better option because it's easy to implement, sampling makes it low latency and doesn't require pre-calculation. Since 2 requires many materializated tables, it may be hard to maintain those tables if you have hundreds of fields in hundreds of collections. I do not have ETA for this issue because in fact you're the first one that wants this feature and this is an enhancement rather than a bug or UI issue. I need your help to implement this feature as we discussed in Slack.

**Comment by sambitdixit** [comment]
Option 1 seems to be better option.

**Comment by rajivonat** [comment]
@kazim-i we need this feature as well.

**Comment by buremba** [comment]
We talked with @saloni2409 about possible implementations, my initial idea was to suggest the values using extra date range. In funnel, retention and event explorer page, the autocomplete could use the active date period. However; we found out that people usually select the date range after they add the filter so this method may not be effective and helpful. For the first implementation, these are the priorities:

1. Since the whole idea of autosuggestion is to help user to enter the values, it must provide low-latency access.
2. We should show most-common values when the user focuses on the input.
3. The values should be consistent and should not require an external service such as Redis, Memcache etc.

Here are some of the ideas for implementation:

1. We can dynamically execute `SELECT DISTINCT mobile_platform FROM collection WHERE prefix LIKE 'and%' WHERE _time BETWEEN now() - INTERVAL '30' and now() LIMIT 50`. It doesn't require materialization and for big data-sets, we can basically use TABLE SAMPLE feature in order to make sure that latency is low.
```
Pros: The result will be accurate and the implementation doesn't require much work.
Cons: It's hard to provide low-latency and if the table is huge, `TABLESAMPLE` may not be consistent and return the expected values. (I believe that we can make it consistent though)
```

2. We can use materialized tables to store the most common 1M values for a given field for the last 30 days. For each day, we can insert the unique values into a table and prune the last day from the table. 
`CREATE TABLE distinct_field_values (collection VARCHAR, attribute VARCHAR, value VARCHAR, seen_at DATE);`

Initial creation:
```
INSERT INTO distinct_field_values AS SELECT 'collection', attribute, now() FROM collection WHERE _time > now() - INTERVAL '30' days LIMIT 1000000
```

On update:
```
INSERT INTO distinct_field_values AS SELECT 'collection', attribute, now() FROM collection WHERE _time > now() - INTERVAL '1' days  LIMIT 100000;
DELETE FROM distinct_field_values WHERE _time < now() - INTERVAL '30' days
```

Query data:
```
SELECT value distinct_field_values FROM  WHERE collection = 'collection' and value LIKE '%android%' LIMIT 50
```
The query should run under a second ideally, if it doesn't work we can even fetch the ORC files from S3 using API since the data is not big, stream the data and find 50 items. 

```
Pros: Low-latency guarantee, no extra dependency and accurate enough since we won't use sampling.
Cons: A little bit complex compared to the first implementation and not real-time (daily).
```
Let me know what you think @saloni2409
